apps/prairielearn/src/sync/syncFromDisk.sql

```diff
@@ -0,0 +1,18 @@
+-- BLOCK select_shared_questions
```

Comment:
I think there's a more efficient version of this query (please confirm that this is correct though):

```sql
SELECT
  q.id,
  q.qid
FROM
  questions AS q
WHERE
  q.course_id = $course_id
  AND (
    q.shared_publicly
    OR EXISTS (
      SELECT
        1
      FROM
        sharing_set_questions
      WHERE
        question_id = q.id
    )
  );
```

Here's the query plan for the above when executed against an actual course with shared questions in production:

```
                                                                  QUERY PLAN                                                                  
----------------------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on questions q  (cost=1439.55..16408.94 rows=997 width=37) (actual time=2.772..2.773 rows=0 loops=1)
   Recheck Cond: (course_id = 75)
   Filter: (shared_publicly OR (hashed SubPlan 2))
   Rows Removed by Filter: 1984
   Heap Blocks: exact=531
   ->  Bitmap Index Scan on questions_course_id_uuid_key  (cost=0.00..1439.31 rows=1985 width=0) (actual time=0.792..0.792 rows=1984 loops=1)
         Index Cond: (course_id = 75)
   SubPlan 2
     ->  Seq Scan on sharing_set_questions  (cost=0.00..6.45 rows=345 width=8) (actual time=0.006..0.047 rows=349 loops=1)
 Planning Time: 0.154 ms
 Execution Time: 2.825 ms
(11 rows)
```

Compare that to the original query:

```
                                                                         QUERY PLAN                                                                          
-------------------------------------------------------------------------------------------------------------------------------------------------------------
 HashAggregate  (cost=11034.99..11035.14 rows=15 width=40) (actual time=3.752..3.754 rows=0 loops=1)
   Group Key: questions.id, questions.qid
   Batches: 1  Memory Usage: 24kB
   ->  Append  (cost=1439.31..11034.92 rows=15 width=40) (actual time=3.750..3.752 rows=0 loops=1)
         ->  Bitmap Heap Scan on questions  (cost=1439.31..8141.17 rows=9 width=37) (actual time=2.459..2.459 rows=0 loops=1)
               Recheck Cond: (course_id = 75)
               Filter: shared_publicly
               Rows Removed by Filter: 1984
               Heap Blocks: exact=531
               ->  Bitmap Index Scan on questions_course_id_uuid_key  (cost=0.00..1439.31 rows=1985 width=0) (actual time=0.786..0.786 rows=1984 loops=1)
                     Index Cond: (course_id = 75)
         ->  Unique  (cost=2893.47..2893.52 rows=6 width=37) (actual time=1.290..1.290 rows=0 loops=1)
               ->  Sort  (cost=2893.47..2893.49 rows=6 width=37) (actual time=1.289..1.290 rows=0 loops=1)
                     Sort Key: q.id, q.qid
                     Sort Method: quicksort  Memory: 25kB
                     ->  Nested Loop  (cost=0.42..2893.40 rows=6 width=37) (actual time=1.282..1.282 rows=0 loops=1)
                           ->  Seq Scan on sharing_set_questions ssq  (cost=0.00..6.45 rows=345 width=8) (actual time=0.007..0.036 rows=349 loops=1)
                           ->  Index Scan using questions_pkey on questions q  (cost=0.42..8.37 rows=1 width=37) (actual time=0.003..0.003 rows=0 loops=349)
                                 Index Cond: (id = ssq.question_id)
                                 Filter: (course_id = 75)
                                 Rows Removed by Filter: 1
 Planning Time: 0.312 ms
 Execution Time: 3.815 ms
(23 rows)
```

I acknowledge that the speed gains are pretty small, but I think it's worth it?

Reply:
changed, thanks!